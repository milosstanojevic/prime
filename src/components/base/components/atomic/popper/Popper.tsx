import React, {
    forwardRef,
    useCallback,
    useRef,
    useEffect,
    useState,
    CSSProperties,
    HTMLAttributes
} from 'react';
import { createPopper, Instance, State } from '@popperjs/core';
import { Placement } from './Placement';
import useForkRef from '../../../hooks/useForkRef';

interface TransitionProps {
    in: boolean;
    onEnter: () => void;
    onExited: () => void;
}
interface IChildProps {
    placement: Placement;
    transitionProps: Partial<TransitionProps>;
}

interface PopperProps {
    /** Anchor Element. */
    anchorEl?: Element | null;
    /** Open */
    open?: boolean;
    /** Inline style */
    style?: CSSProperties;
    /** Child with transition */
    transition?: boolean;
    /** Placement */
    placement?: Placement;
}

export const Popper = forwardRef<HTMLDivElement, PopperProps & HTMLAttributes<HTMLDivElement>>(
    (
        {
            children,
            anchorEl,
            open = false,
            style,
            transition = false,
            placement: initialPlacement = 'bottom',
            ...rest
        },
        ref
    ) => {
        const [exited, setExited] = useState(true);
        const [placement, setPlacement] = useState(initialPlacement);
        const ownRef = useRef<HTMLDivElement | null>(null);
        const popperRef = useRef<Instance>();

        const handleInnerRef = useForkRef(ownRef, ref);

        const handlePopperUpdate = (data: Partial<State>): void => {
            setPlacement(data.placement || 'bottom');
        };

        const handleOpen = useCallback(() => {
            if (!ownRef.current || !anchorEl || !open) {
                return;
            }

            if (popperRef.current) {
                popperRef.current.destroy();
            }

            popperRef.current = createPopper(anchorEl, ownRef.current, {
                placement,
                onFirstUpdate: handlePopperUpdate
            });
        }, [anchorEl, open, placement]);

        const handleClose = (): void => {
            if (!popperRef.current) {
                return;
            }

            popperRef.current.destroy();
        };

        const handleRef = useCallback(
            (node: HTMLDivElement) => {
                handleInnerRef(node);
                handleOpen();
            },
            [handleOpen, handleInnerRef]
        );

        const handleEnter = (): void => {
            setExited(false);
        };

        const handleExited = (): void => {
            setExited(true);
            handleClose();
        };

        useEffect(() => {
            return (): void => {
                handleClose();
            };
        }, []);

        useEffect(() => {
            if (!open && !transition) {
                handleClose();
            }
        }, [open, transition]);

        useEffect(() => {
            if (popperRef.current) {
                popperRef.current.update();
            }
        });

        if (!open && (!transition || exited)) {
            return null;
        }

        const childProps: Partial<IChildProps> = { placement };

        if (transition) {
            childProps.transitionProps = {
                in: open,
                onEnter: handleEnter,
                onExited: handleExited
            };
        }

        return (
            <div
                ref={handleRef}
                {...rest}
                style={{
                    ...style
                }}
            >
                {typeof children === 'function'
                    ? React.cloneElement(children, [childProps])
                    : children}
            </div>
        );
    }
);

Popper.displayName = 'Popper';
