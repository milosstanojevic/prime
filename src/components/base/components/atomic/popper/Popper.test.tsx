import React, {ReactElement} from "react";
import { mount, shallow } from "enzyme";
import { Popper } from "./Popper";
const mockUpdate = jest.fn();
const mockDestroy = jest.fn();

jest.mock("@popperjs/core", () => ({
  createPopper: jest.fn().mockImplementation(() => ({
    update: mockUpdate,
    destroy: mockDestroy,
  })),
}));

describe("<Popper/>", () => {

  afterEach(() => {
    mockUpdate.mockReset();
    mockDestroy.mockReset();
  })

  it("should render", () => {
    const wrapper = shallow(
      <Popper open transition>
        <div id="div-element">Div</div>
      </Popper>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should not render because it's exited", () => {
    const wrapper = shallow(
      <Popper open>
        <div id="div-element">Div</div>
      </Popper>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should not render", () => {
    const wrapper = shallow(
      <Popper open={false}>
        <div id="div-element">Div</div>
      </Popper>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should not render with transition on", () => {
    const wrapper = shallow(
      <Popper open={false} transition>
        <div id="div-element">Div</div>
      </Popper>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should render without transition", () => {
    const wrapper = shallow(
      <Popper open>
        <div id="div-element">Div</div>
      </Popper>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should render with props", () => {
    const wrapper = shallow(
      <Popper
        open
        style={{ color: 'red', }}
        placement="right"
      >
        {(): ReactElement => (
          <div id="div-element">Div</div>
        )}
      </Popper>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("should mount", () => {
    const ref = React.createRef<HTMLDivElement>();
    const wrapper = mount(
      <Popper
        open
        transition
        ref={ref}
      >
        {(props: { transitionProps: { onEnter: ((event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void) | undefined; onExited: ((event: React.MouseEvent<HTMLDivElement, MouseEvent>) => void) | undefined; }; }): ReactElement => {
          return (
            <>
              <div id="div-element-enter" onClick={props.transitionProps.onEnter}>OnEnter</div>
              <div id="div-element-exit" onClick={props.transitionProps.onExited}>OnExited</div>
            </>
          )
        }}
      </Popper>
    );
    wrapper.find('#div-element-enter').simulate('click');
    expect(wrapper.find('#div-element-enter').text()).toBe('OnEnter');
    wrapper.find('#div-element-exit').simulate('click');
    expect(wrapper.find('#div-element-exit').text()).toBe('OnExited');
    wrapper.unmount();
  });

  it("should create popper instance", () => {
    const node = document.getElementsByTagName('body')[0];
    const wrapper = mount(
      <Popper
        open
        transition
        anchorEl={node}
      >
        <span>Test</span>
      </Popper>
    );
    expect(mockDestroy).not.toHaveBeenCalled();
    wrapper.unmount();
    expect(mockDestroy).toHaveBeenCalled();
  });

  it("should handle close when props has been changed", () => {
    const node = document.getElementsByTagName('body')[0];
    const wrapper = mount(
      <Popper open anchorEl={node}>
        <span>Test</span>
      </Popper>
    );
    expect(mockUpdate).toHaveBeenCalledTimes(1);
    expect(mockDestroy).not.toHaveBeenCalled();
    wrapper.setProps({ open: false });
    expect(mockUpdate).toHaveBeenCalledTimes(2);
    expect(mockDestroy).toHaveBeenCalled();
    wrapper.unmount();
  });
});
