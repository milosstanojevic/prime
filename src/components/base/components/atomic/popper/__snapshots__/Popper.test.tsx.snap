// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Popper/> should not render 1`] = `ShallowWrapper {}`;

exports[`<Popper/> should not render because it's exited 1`] = `ShallowWrapper {}`;

exports[`<Popper/> should not render with transition on 1`] = `ShallowWrapper {}`;

exports[`<Popper/> should render 1`] = `ShallowWrapper {}`;

exports[`<Popper/> should render with props 1`] = `ShallowWrapper {}`;

exports[`<Popper/> should render without transition 1`] = `ShallowWrapper {}`;
